crates/polars-ops/src/frame/join/hash_join/single_keys.rs

```diff
@@ -1,53 +1,146 @@
     POOL.install(|| {
-        (0..n_partitions)
-            .into_par_iter()
-            .map(|partition_no| {
-                let partition_no = partition_no as u64;
+        // Compute the number of elements in each partition for each portion.
+        let per_thread_partition_sizes: Vec<Vec<usize>> = keys
+            .par_iter()
+            .map(|key_portion| {
+                let mut partition_sizes = vec![0; n_partitions];
+                for key in key_portion.clone() {
+                    let h = key.as_u64();
+                    let p = hash_to_partition(h, n_partitions);
+                    unsafe {
+                        *partition_sizes.get_unchecked_mut(p) += 1;
+                    }
+                }
+                partition_sizes
+            })
+            .collect();
 
-                let mut hash_tbl: PlHashMap<T, Vec<IdxSize>> =
-                    PlHashMap::with_capacity(_HASHMAP_INIT_SIZE);
+        // Compute output offsets with a cumulative sum.
+        let mut per_thread_partition_offsets = vec![0; n_partitions * n_threads + 1];
+        let mut partition_offsets = vec![0; n_partitions + 1];
+        let mut cum_offset = 0;
+        for p in 0..n_partitions {
+            partition_offsets[p] = cum_offset;
+            for t in 0..n_threads {
+                per_thread_partition_offsets[t * n_partitions + p] = cum_offset;
+                cum_offset += per_thread_partition_sizes[t][p];
+            }
+        }
+        let num_keys = cum_offset;
+        per_thread_partition_offsets[n_threads * n_partitions] = num_keys;
+        partition_offsets[n_partitions] = num_keys;
 
-                let n_partitions = n_partitions as u64;
-                let mut offset = 0;
-                for keys in &keys {
-                    let keys = keys.clone().into_iter();
-                    let len = keys.size_hint().1.unwrap() as IdxSize;
+        // FIXME: we wouldn't need this if we changed our interface to split the
+        // input in this function, instead of taking a vec of iterators.
+        let mut per_thread_input_offsets = vec![0; n_partitions];
+        cum_offset = 0;
+        for t in 0..n_threads {
+            per_thread_input_offsets[t] = cum_offset;
+            for p in 0..n_partitions {
+                cum_offset += per_thread_partition_sizes[t][p];
+            }
+        }
 
-                    let mut cnt = 0;
-                    keys.for_each(|k| {
-                        let idx = cnt + offset;
-                        cnt += 1;
+        // Scatter values into partitions.
+        let mut scatter_keys: Vec<T> = Vec::with_capacity(num_keys);
+        let mut scatter_idxs: Vec<IdxSize> = Vec::with_capacity(num_keys);
+        let scatter_keys_ptr = unsafe { SyncPtr::new(scatter_keys.as_mut_ptr()) };
+        let scatter_idxs_ptr = unsafe { SyncPtr::new(scatter_idxs.as_mut_ptr()) };
+        keys.into_par_iter()
```

Comment:
I understand this is the gist of it. Any ideas what the chances of false sharing are here? In proper hashing very low I assume.